File: expert_system/dish_recomendation/dish_recomendation.py
class SistemaEspecialista:
    def __init__(self):
        self.pratos = [
            {
                "nome": "Salada Vegetariana",
                "tipos": ["vegetariano", "salada", "salgado"],
                "restriçoes_prato": ["verduras", "hortalicas"],
            },
            {
                "nome": "Lasanha",
                "tipos": ["massa", "carne", "salgado"],
                "restriçoes_prato": ["massa", "ovo", "carne"],
            },
            {
                "nome": "Sushi",
                "tipos": ["peixe", "vegetariano", "salgado"],
                "restriçoes_prato": ["frutos do mar"],
            },
            {
                "nome": "Frango Assado",
                "tipos": ["carne", "salgado"],
                "restriçoes_prato": ["carne"],
            },
            {
                "nome": "Sorvete de Chocolate",
                "tipos": ["doce", "vegetariano"],
                "restriçoes_prato": ["lactose"],
            },
        ]

    def verificar_preferencias(self, prato, cliente):
        return any(tipo in cliente["preferencias"] for tipo in prato["tipos"])

    def verificar_restricoes(self, prato, cliente):
        return not any(
            restricoes in cliente["restricoes"]
            for restricoes in prato["restriçoes_prato"]
        )

    def recomendar_prato(self, cliente):
        pratos_compativeis = []
        for prato in self.pratos:
            if self.verificar_preferencias(
                prato, cliente
            ) and self.verificar_restricoes(prato, cliente):
                pratos_compativeis.append(prato["nome"])
        return pratos_compativeis


cliente1 = {
    "nome": "João",
    "preferencias": ["vegetariano", "doce"],
    "restricoes": ["lactose"],
}

sistema = SistemaEspecialista()
pratos_recomendados = sistema.recomendar_prato(cliente1)
print(f"Pratos recomendados para {cliente1['nome']}: {pratos_recomendados}")

File: expert_system/dish_recomendation/dish_recomendation.pl
cliente(joao).

prato(salada_vegetariana).
prato(lasanha).
prato(sushi).
prato(frango_assado).
prato(sorvete_de_chocolate).

preferencia(salgado).
preferencia(doce).
preferencia(vegetariano).
preferencia(sobremesa).
preferencia(carne).
preferencia(massa).
preferencia(fruto_do_mar).

restricao(lactose).
restricao(fruto_do_mar).
restricao(carne).
restricao(verdura).
restricao(massa).

recomendar(Cliente, salada_vegetariana) :-
    preferencia_cliente(Cliente, vegetariano),
    preferencia_cliente(Cliente, salgado),
    restricao_cliente(Cliente, verdura).

recomendar(Cliente, lasanha) :-
    preferencia_cliente(Cliente, massa),
    preferencia_cliente(Cliente, salgado),
    restricao_cliente(Cliente, massa).

recomendar(Cliente, sushi) :-
    preferencia_cliente(Cliente, fruto_do_mar),
    preferencia_cliente(Cliente, salgado),
    restricao_cliente(Cliente, fruto_do_mar).

recomendar(Cliente, frango_assado) :-
    preferencia_cliente(Cliente, carne),
    preferencia_cliente(Cliente, salgado),
    restricao_cliente(Cliente, carne).

recomendar(Cliente, sorvete_de_chocolate) :-
    preferencia_cliente(Cliente, sobremesa),
    preferencia_cliente(Cliente, doce),
    restricao_cliente(Cliente, lactose).

preferencia_cliente(joao, vegetariano).
preferencia_cliente(joao, salgado).
restricao_cliente(joao, lactose).